src/routes/DigitalAssets/routes/AddAsset/sagas/addAsset.js

```diff
@@ -0,0 +1,263 @@
+// @flow
+
+import Promise from 'bluebird'
+import keystore from '@jibrelnetwork/jwallet-web-keystore'
+import { put, select, takeEvery, take, race, call, all } from 'redux-saga/effects'
+
+import InvalidFieldError from 'utils/errors/InvalidFieldError'
+
+import {
+  selectAddAsset,
+  selectDigitalAsset,
+} from 'store/stateSelectors'
+
+import { backOrFallback } from 'routes/modules'
+
+import web3 from 'services/web3'
+
+import {
+  OPEN_VIEW,
+  CLOSE_VIEW,
+  SET_FIELD,
+  START_ASSET_LOADING,
+  clean,
+  setField,
+  setFieldError,
+  clearFieldError,
+  setIsAssetValid,
+  startAssetLoading,
+  SUBMIT_ASSET_FORM,
+} from '../modules/addAsset'
+
+import {
+  addCustomAsset,
+} from '../../../modules/digitalAssets'
+
+const {
+  getContractName,
+  getContractSymbol,
+  getContractDecimals,
+  getContractCode,
+  checkContractCodeIsERC20,
+} = web3
+
+type RequestedAssetFields = {|
+  decimals: ?number,
+  name: ?string,
+  symbol: ?string,
+  isERC20: ?boolean,
+|}
+
+const NODE_EMPTY_VALUE = '0x'
+
+/**
+ * Request not required field from contract
+ *
+ * @param {Function} contractMethod
+ * @param {Address} contractAddress
+ */
+function* requestAssetField(
+  contractMethod: Function,
+  contractAddress: Address
+): Saga<string | number | void> {
+  try {
+    const result = yield call(contractMethod, contractAddress)
+    if (result === NODE_EMPTY_VALUE) {
+      return null
+    } else {
+      return result
+    }
+  } catch (err) {
+    // catch Web3 network error
+    if (err instanceof Promise.TimeoutError ||
+      (err.isOperational && err.cause.name !== 'BigNumber Error')) {
+      throw err
+    } else {
+      return null
+    }
+  }
+}
+
+function* checkAssetIsERC20Compatible(contractAddress: Address): Saga<boolean> {
+  const contractCode: string = yield call(getContractCode, contractAddress)
+  return checkContractCodeIsERC20(contractCode)
+}
+
+function* clearFields(): Saga<void> {
+  yield put(setField('name', ''))
+  yield put(setField('symbol', ''))
+  yield put(setField('decimals', ''))
+}
+
+function* clearFieldsError(): Saga<void> {
+  yield put(clearFieldError('address'))
+  yield put(clearFieldError('name'))
+  yield put(clearFieldError('symbol'))
+  yield put(clearFieldError('decimals'))
+}
+
+/**
+ * Fires, when user changes fields on CustomAssetForm
+ *
+ * @param action from setField method
+ * @returns {undefined}
```

Comment:
> @returns {undefined}

If we decide to left such comments, we should write it more correctly.
It returns generator that resolves with undefined